### PR TITLE
 WEA-23 Implement Repository pattern for geocoding api 

### PR DESCRIPTION
**Pull request by ayush-shta**
@subash-poudel

### CODE REVIEW

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java [hunk 1]
```diff
@@ -168,27 +164,38 @@ private void stopLocationUpdates() {
   }
 
   private void fetchAddress(String latLng) {
+    // TODO refactor after mvp complete
+    Disposable disposable = geocodingRepository.getLocationDetails(latLng)
+        .subscribeOn(Schedulers.io())
+        .observeOn(AndroidSchedulers.mainThread())
+        .subscribeWith(getGeoLocationObserver());
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: Don't copy paste code from DT. Creating the function serves no useful purpose here.

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java [hunk 2]
```diff
@@ -168,27 +164,38 @@ private void stopLocationUpdates() {
...
+    };
+  }
+
+  private void setLocation(GeoLocation geoLocation) {
+    List<Address> addressList = geoLocation.getAddresses();
+
+    if (addressList != null && !addressList.isEmpty()) {
+      List<AddressComponents> addressComponentsList =
+          addressList.get(0).getAddressComponents();
+      getView().setAddress(getAddress(addressComponentsList));
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: getAddress function can check for null and empty and return the string from else block.

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java [hunk 3]
```diff
@@ -255,22 +262,26 @@ private void fetchWeatherForecast(String latLng) {
...
-            getView().changeErrorVisibility(true);
-            getView().showErrorMessage();
-            getView().hideProgressBar();
-          }
-
-          @Override public void onComplete() {
-            getView().hideProgressBar();
-          }
-        });
+        .subscribeWith(getWeatherForecastObserver(latLng));
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: redundant function

File: app/src/main/java/com/ayush/weatherapp/repository/geocoding/GeocodingRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+  private GeocodingAPIInterface geocodingAPIInterface;
+
+  // TODO provide dependencies using dagger
+  public GeocodingRepositoryImpl() {
+    geocodingAPIInterface = GeocodingAPIClient.getClient().create(GeocodingAPIInterface.class);
+  }
+
+  @Override public Observable<GeoLocation> getLocationDetails(String latlng) {
+    return geocodingAPIInterface.getLocationDetailsRX(latlng);
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/repository/geocoding/GeocodingRepositoryImpl.java]: remove RX postfix

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java (hunk 1 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java (hunk 2 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/repository/geocoding/GeocodingRepositoryImpl.java (hunk 4 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/repository/geocoding/GeocodingRepositoryImpl.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java (hunk 3 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: :ok:

**Pull request merged**